lib/mdw_fused_dslash.cu

```diff
@@ -0,0 +1,757 @@
+      // tp.aux.x dictates whether or not to reload.
+      template <int block_dim_x, bool reload, int type>
+      void apply(const TuneParam &tp, Arg &arg, const cudaStream_t &stream)
+      {
+        switch (tp.aux.y) {
+        case 1: launch(fused_tensor_core<block_dim_x, 1, reload, Arg, type>, tp, arg, stream); break;
+        case 2: launch(fused_tensor_core<block_dim_x, 2, reload, Arg, type>, tp, arg, stream); break;
+        case 3: launch(fused_tensor_core<block_dim_x, 3, reload, Arg, type>, tp, arg, stream); break;
+        default: errorQuda("NOT valid tp.aux.y(=%d)\n", tp.aux.y);
+        }
+      }
+
+      template <bool reload, int type> void apply(const TuneParam &tp, Arg &arg, const cudaStream_t &stream)
+      {
+        switch (tp.block.x) {
+        case 16: apply<16, reload, type>(tp, arg, stream); break;
+        case 32: apply<32, reload, type>(tp, arg, stream); break;
+        default: errorQuda("NOT valid tp.block.x(=%d)\n", tp.block.x);
+        }
+      }
+
+      template <int type> void apply(const TuneParam &tp, Arg &arg, const cudaStream_t &stream)
+      {
+        if (tp.aux.x == 0) {
+          apply<false, type>(tp, arg, stream); // reload = false
+        } else {
+          apply<true, type>(tp, arg, stream); // reload = true
+        }
+      }
+
+      void apply(const cudaStream_t &stream)
+      {
+        TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+        switch (arg.type) {
+        case 0: apply<0>(tp, arg, stream); break;
+        case 1: apply<1>(tp, arg, stream); break;
+        case 2: apply<2>(tp, arg, stream); break;
+        case 3: apply<3>(tp, arg, stream); break;
+        case 4: apply<4>(tp, arg, stream); break;
+        default: errorQuda("Unknown MdwfFusedDslashType %d", arg.type);
+        }
+      }
+
+      void initTuneParam(TuneParam &param) const
+      {
+        Tunable::initTuneParam(param);
+        param.block = dim3(blockMin(), arg.Ls, 1); // Ls must be contained in the block
+        param.grid = dim3(minGridSize(), 1, 1);
+        param.shared_bytes = sharedBytesPerBlock(param);
+        param.aux.x = 0;
+        param.aux.y = 1;
+      }
+
+      void defaultTuneParam(TuneParam &param) const { initTuneParam(param); }
+
+      TuneKey tuneKey() const { return TuneKey(meta.VolString(), typeid(*this).name(), aux); }
+    };
+
+    // Apply the 5th dimension dslash operator to a colorspinor field
+    // out = Dslash5 * in
+    template <typename storage_type, int nColor, QudaReconstructType recon> struct FusedApply {
+
+      inline FusedApply(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, ColorSpinorField &y,
+                        const ColorSpinorField &x, double m_f, double m_5, const Complex *b_5, const Complex *c_5,
+                        bool dagger, int parity, int shift[4], int halo_shift[4], MdwfFusedDslashType type)
+      {
+        // switch for Ls
+        switch (in.X(4)) {
+        case 4: {
```

Comment:
I imagine the factors of 4 are due to the tensor core sizes, can a quick comment on this get added?